src/resource/bindless.rs

```diff
@@ -0,0 +1,243 @@
+                    s_type: vk::StructureType::WRITE_DESCRIPTOR_SET,
+                    p_next: std::ptr::null(),
+                    dst_set: self.descriptor_set.handle,
+                    dst_binding: 0,
+                    dst_array_element: i,
+                    descriptor_count: 1,
+                    descriptor_type: R::descriptor_type(),
+                    p_image_info: &r.descriptor_info() as *const _,
+                    p_buffer_info: std::ptr::null(),
+                    p_texel_buffer_view: std::ptr::null(),
+                }
+            })
+            .collect::<Vec<_>>();
+
+        unsafe {
+            self.descriptor_set.device.update_descriptor_sets(vk_writes.as_slice(), &[]);
+        }
+    }
+
+
+    fn take(&mut self, key: u32) -> Option<R> {
+        if key <= self.items.len() as _ {
+            None
+        } else {
+            self.items[key as usize].take().and_then(|ob| {
+                self.free.push(key);
+                Some(ob)
+            })
+        }
+    }
+}
+
+/// A bindless pool can hold a number of resources.
+/// It will keep a descriptor set up to date so that any resource contained in it.
+pub struct BindlessPool<R> {
+    inner: Arc<Mutex<BindlessPoolInner<R>>>,
+}
+
+impl<P: BindlessResource> BindlessPool<P> {
+    fn with<F: FnOnce(&mut BindlessPoolInner<P>) -> R, R>(&self, f: F) -> R {
+        let mut inner = self.inner.lock().unwrap();
+        f(&mut inner)
+    }
+
+    /// Allocate a single item from the pool.
+    pub fn alloc(&self, item: P) -> BindlessHandle<P> {
+        self.with(|p| {
+            let key = p.free
+                .pop()
+                .unwrap_or_else(|| {
+                    p.items.push(None);
+                    p.items.len() as u32 - 1
+                });
+            p.update_descriptor_set(std::iter::once((key, &item)));
+            p.items[key as usize] = Some(item);
+            BindlessHandle {
+                key,
+                pool: Self { inner: self.inner.clone() }
+            }
+        })
+    }
+
+    /// Allocate a number of items from the pool.
+    pub fn alloc_items(&self, items: &[P]) -> impl Iterator<Item = BindlessHandle<P>> {
+        self.with(|p| {
+            let mut keys = p.free.iter().cloned().rev().take(items.len()).collect::<Vec<_>>();
+            if keys.len() < items.len() {
+                let old_len = p.items.len() as u32;
+                p.items.resize_with(items.len(), || None);
+                keys.extend(old_len..p.items.len() as u32);
+            }
+            p.update_descriptor_set(keys.iter().cloned().zip(items));
+            let pool_inner = self.inner.clone();
+            keys
+                .into_iter()
+                .map(move |key| {
+                    BindlessHandle {
+                        key,
+                        pool: Self { inner: pool_inner.clone() }
+                    }
+                })
+        })
+    }
+}
+
+impl<P: BindlessResource> BindlessPool<P> {
```

Comment:
Is there a reason this is in a different `impl` block

Reply:
They probably differed at some point but clearly no point in keeping it that way now